server/src/main/java/org/elasticsearch/action/support/TransportAction.java

```diff
@@ -54,6 +56,13 @@ public final void execute(Task task, Request request, ActionListener<Response> l
         requestFilterChain.proceed(task, actionName, request, listener);
     }
 
+    /**
+     * Prepare the request before executing it. Defaults to no-op.
+     */
+    protected void prepareRequest(Request request) {
```

Comment:
I tried to make this an ActionFilter, but we need a hack to do this. The new action filter must be executed before the security action filter, but the security action filter already has the lowest order (Integer.MIN_VALUE).

Reply:
I see the problem Nhat, thanks for raising this. Adding a new base method that runs before action filters makes me nervous. Possibly an alternative hack: would it work to make what SearchReqest#indices returns dynamic and extract the indices from pit if a pit is set? In that case the authorization filter would look at the right indices and we would not need to set the indices manually? I guess there can't be wildcards in there hence the authorization filter will not replace the indices in that scenario. there may be other side-effects though around going down this route.

Reply:
I see the problem Nhat, thanks for raising this. Adding a new base method that runs before action filters makes me nervous. Possibly an alternative hack: would it work to make what SearchReqest#indices returns dynamic and extract the indices from pit if a pit is set? In that case the authorization filter would look at the right indices and we would not need to set the indices manually? I guess there can't be wildcards in there hence the authorization filter will not replace the indices in that scenario. there may be other side-effects though around going down this route.

Reply:
@ywangd what do you think?

Reply:
If we make `SearchRequest#indices()` method dynamic, how would the corresponding setter behave? `SearchRequest` implements `IndicesRequest.Replaceable`. The contract for authorization code is that it will always call `SearchRequest#indices(String ..)` with the expanded indices regardless whether the expanded indices is identical to the original one. 

IIUC, in practice, Search with PIT always gets back to the identical list of indices from Authorization because it does not allow `ignore_unavailable=true`. So technically the setter can be a no-op, but I am not really comfortable with that because it feels PIT has to know how Authorization works. It also excludes the possibility to ever support `ignore_unavailable=true`. One option is to make the getter dynamically return the list indices from PIT if setter is never called, but otherwise return the value configured by the setter method (maybe this is already what @javanna suggests by "dynamic"?). The downside is inconsistency between the `indices()` method and the internal `indices` variable (till the setter is called).

I am ok with what @dnhatn has proposed. Though I'd rather add the new base method to `ActionRequest` instead of `TransportAction` because calling `ActionRequest#validate()` and `ActionRequest#prepare()` in sequence seems to be more fluent to me (it would be nice to make `ActionRequest#prepare` idempotent). 

Overall, I slightly prefer the `TransportAction` (or `ActionRequest`) approach which feels simpler.

Reply:
as far as I understand the pit indices never hold wildcards, and never will, because you need to first open a pit on a fixed set of indices. So the whole indices expansion and replacement does not really apply to this specific scenario. It would be great to have a different request for this scenario, but we don't unfortunately. The indices variable is only read by methods within SearchRequest that could call the indices() method instead for consistency? We could even say that the setter throws an exception in case a pit is set.

Reply:
Indices expansion in authorization does not have knowledge about PIT. Its behaviour is largely determined by whether the incoming request is a `IndicesRequest.Replaceable`. If we could turn SearchRequest with PIT into a new class that does not implementing this interface, things should just work on the authorization side. I think this would be better than throwing exception in the setter method.

Reply:
> Indices expansion in authorization does not have knowledge about PIT.

Correct. What I am saying is that because you need to explicitly open a PIT with a specific API call, the PIT id will always hold concrete indices and never wildcards when set to a search request, because you previously received it as part of the open pit response, so despite the search request implements IndicesRequest.Replaceable, there will never be anything to expand when a PIT id is set. You were reasoning about ignore_unavailable and wildcards expansion in your previous reply, but these don't apply to a search when a PIT id is set to it. I think this is important as it reduces the scope of how indices are resolved and should be authorized in this specific scenario. I believe this is key to finding a simple solution for the current problem. I do not think that introducing a new base method to TransportAction for this very specific edge-case is a good trade-off.

> If we could turn SearchRequest with PIT into a new class that does not implementing this interface, things should just work on the authorization side. I think this would be better than throwing exception in the setter method.

That would make things better, and maybe it is possible given that we have control over the REST action, though it looks to me as it has quite a cost in terms of added complexity and lines of code. We would need to try that to see it in reality.

Reply:
> though it looks to me as it has quite a cost in terms of added complexity and lines of code.

Yeah Java does not make delegation easy. 

> You were reasoning about ignore_unavailable and wildcards expansion in your previous reply, but these don't apply to a search when a PIT id is set to it.

Right I understand PIT does not really need expansion. My previous reply unnecessarily blurred the focus. My point is Authorization code will call `Replaceable#indices(String...)` as long as the request is an instance of `Replaceable`. I don't particularly like having a special setter method for `SearchRequest` because it indicates SearchRequest knows about authorization. But I am ok with the opposite, i.e. Authorization knows more about SearchRequest because it already knows quite a bit about the requests. The following is where Authorization decides whether to expand wildcards (and in turn calls the setter method):

https://github.com/elastic/elasticsearch/blob/11599e1e5f269679f3b96da76760ebd90610ab1e/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L131-L141

We can add an addition check for PIT before `if (indicesRequest instanceof IndicesRequest.Replaceable)` such as:
```java
if (indicesRequest instanceof final SearchRequest searchRequest && searchRequest.pointInTimeBuilder() != null) {
    return false;
}
```

The above check in conjunction with a dynamic `SearchRequest#indices()` getter method should work and the change is rather minimal. What do you think?

Reply:
@javanna @ywangd Thanks for the discussion.

> The above check in conjunction with a dynamic SearchRequest#indices() getter method should work and the change is rather minimal. What do you think?

This works for me.